File: Lab3/src/main/java/ubb/scs/map/repository/memory/InMemoryRepository.java
package ubb.scs.map.repository.memory;


import ubb.scs.map.domain.Entity;
import ubb.scs.map.domain.validators.ValidationException;
import ubb.scs.map.domain.validators.Validator;
import ubb.scs.map.repository.Repository;

import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID, E> {

    private final Validator<E> validator;
    protected Map<ID, E> entities;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities = new HashMap<ID, E>();
    }

    @Override
    public E findOne(ID id) {
        return entities.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) throws ValidationException {
        if (entity == null) throw new IllegalArgumentException("ENTITY CANNOT BE NULL");
        validator.validate(entity);
        if (entities.containsKey(entity.getId())) return entity;
        else {
            entities.put(entity.getId(), entity);
            return null;
        }
    }

    @Override
    public E delete(ID id) {
        if (id == null) throw new IllegalArgumentException("ID CANNOT BE NULL");
        return entities.remove(id);
    }

    @Override
    public E update(E entity) {
        if (entity == null) throw new IllegalArgumentException("ENTITY CANNOT BE NULL");
        validator.validate(entity);
        if (entities.containsKey(entity.getId())) return entity;
        entities.put(entity.getId(), entity);
        return null;
    }
}

File: Lab10/Lab10/Repository/InMemoryRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lab10.Domain.Validators;
using Lab10.Repository;

namespace Lab10.Domain;

public class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
{
    private readonly IValidator<E> validator;
    protected Dictionary<ID, E> entities;

    public InMemoryRepository(IValidator<E> validator)
    {
        this.validator = validator;
        entities = new Dictionary<ID, E>();
    }

    public E FindOne(ID id)
    {
        if (id == null) throw new ArgumentException("ID CANNOT BE NULL");
        entities.TryGetValue(id, out var entity);
        return entity;
    }

    public IEnumerable<E> FindAll()
    {
        return entities.Values;
    }

    public virtual E Save(E entity)
    {
        if (entity == null) throw new ArgumentException("ENTITY CANNOT BE NULL");
        validator.Validate(entity);
        if (entities.ContainsKey(entity.Id)) return entity;
        entities[entity.Id] = entity;
        return null;
    }

    public virtual E Delete(ID id)
    {
        if (id == null) throw new ArgumentException("ID CANNOT BE NULL");
        if (entities.TryGetValue(id, out var entity))
        {
            entities.Remove(id);
            return entity;
        }

        return null;
    }

    public virtual E Update(E entity)
    {
        if (entity == null) throw new ArgumentException("ENTITY CANNOT BE NULL");
        validator.Validate(entity);
        if (entities.ContainsKey(entity.Id))
        {
            entities[entity.Id] = entity;
            return null;
        }

        return entity;
    }
}